[docs] In progress docs explaining Object versions (#8842)
MystenLabs/sui: doc/in-progress/object_versions.md

Versions are strictly increasing and (ID, version) pairs are never re-used. This structure allows node operators to prune their stores of old object versions that are now inaccessible, if they choose. This is not a requirement, though, as node operators might keep prior object versions around to serve requests for an object's history, either from other nodes that are catching up, or from RPC requests.

## Move Objects

Sui uses [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamp) in its versioning algorithm for objects. The use of Lamport timestamps guarantees that versions never get re-used:  The new version for objects touched by a transaction is one greater than the max version among all input objects to the transaction.  For example, a transaction transferring an object `O` at version `5` using a gas object `G` at version `3` updates both `O` and `G` versions to `1 + max(5, 3) = 6`.

> The relevance of versions for accessing an object as a transaction input changes depending on that object's ownership.

Reviewer: Maybe have a snippet for: "Why this scheme versus something simpler? See the sections below on wrapped objects and dynamic fields"

### Shared Objects

Specifying a shared transaction input is slightly more complex. You reference it by its ID, the version it was shared at, and a flag indicating whether it is accessed mutably. You don't specify the precise version the transaction accesses because it is decided by consensus during transaction scheduling. When scheduling multiple transactions that touch the same shared object, validators agree the order of those transactions, and pick each transaction's input versions for the shared object accordingly (one transaction's output version becomes the next transaction's input version, and so on).

Shared transaction inputs that you reference immutably participate in scheduling, but don't modify the object or increment its version.

### Wrapped Objects

The `make_wrapped` function in the following example creates an `Inner` object, wrapped in an `Outer` object, which is sent back to the transaction sender.

```
module example::wrapped {
    use sui::object::{Self, UID};
    use sui::transfer;
    use sui::tx_context::{Self, TxContext};
    

>     struct Inner has key {

Reviewer: ```suggestion
    struct Inner has key, store {
```

        id: UID,
        x: u64,
    }
    
    struct Outer has key {
        id: UID,
        inner: Inner,
    }
    
    entry fun make_wrapped(ctx: &mut TxContext) {
        transfer::transfer(
            Outer {
                id: object::new(ctx),
                inner: Inner {
                    id: object::new(ctx),
                    x: 42,
                }
            }
            tx_context::sender(ctx),
        )
    }
}
```

> You can't access wrapped objects by their ID in the object store, you must access them by the object that wraps them. In the prior example, the owner of `Outer` must specify it as the transaction input and then access its `inner` field to read the instance of `Inner`. Validators refuse to sign transactions that specify wrapped objects (like the `inner` of an `Outer`) as inputs.  As a result, you don't need to specify a wrapped object's version in a transaction that reads that object.
> 
> Wrapped objects can eventually become "unwrapped", meaning that they are once again accessible at their ID:

Reviewer: ```suggestion
Wrapped objects are like deleted objects in that can't access wrapped objects by their ID in the object store, you must access them by the object that wraps them. In the prior example, the owner of `Outer` must specify it as the transaction input and then access its `inner` field to read the instance of `Inner`. Validators refuse to sign transactions that specify wrapped objects (like the `inner` of an `Outer`) as inputs.  As a result, you don't need to specify a wrapped object's version in a transaction that reads that object.

Unlike deleted objects, wrapped objects can eventually return to the object store, becoming "unwrapped". After this, they are once again accessible at their ID:
```

Reviewer: Thoughts? I'm trying to put in some parallels to deleted objects but I'm not sure if it is actually helpful

- After a transaction, `W` wrapping object `I` in object `O`, `O`'s version is greater than or equal to `I`'s. This means one of the following conditions is true:
  - `I` is an input so has a strictly lower version.
  - `I` is new and has an equal version.
- After a later transaction unwrapping `I` out of `O`, the following must be true: 
  - `O`'s input version is greater than or equal to its version after `W` because it is a later transaction, so the version can only have increased.
  - `I`'s version in the output must be strictly greater than `O`'s input version.
  
This leads to the following chain of inequalities for `I`'s version before wrapping:

- less than or equal to `O`'s version after wrapping
- less than or equal to `O`'s version before unwrapping
- less than `I`'s version after unwrapping

So `I`'s version before wrapping is less than `I`'s version after unwrapping.

### Dynamic Fields

From a versioning perspective, values held in dynamic fields behave like wrapped objects:

- They are only accessible via their field owner, not as direct transaction inputs.

> - Based on the previous point, you do not need to supply their versions with the transaction inputs.

Reviewer: ```suggestion
- Based on the previous point, they are not specified as transaction inputs. In other words, neither the object's ID nor object's version need to be specified ahead of time.
```

> - Lamport timestamp-based versioning makes sure that when a transaction removes a field and its value becomes accessible by its ID, its version has been incremented.

Reviewer: ```suggestion
- In the case that the dynamic field as holding an object, the case is exactly the same as wrapped objects; Lamport timestamp-based versioning makes sure that when a transaction removes a field and its value becomes accessible by its ID, its version has been incremented to a previously unused version.
```

> 

Reviewer: Big thing missing here is that Lamport versioning solves the creation/deletion/creation problem. I tried writing a short blurb about it but its a bit of a mess
```
In order to truly understand what problem Lamport versioning solves for dynamic fields, we need to go a bit more in depth into how dynamic fields are implemented. Dynamic fields are objects themselves, but the owner of the field object is another object. We call this object the parent object. When creating the dynamic field, it's ID is _not_ created based on the Transaction Digest. Instead we hash information that is used to access the field, specifically the ID of the parent, the "name" of the field (a Move value), and the type of the "name" (the type of the Move value). In other words, the ID of a dynamic field object is determined by `hash(parent.id || name || type(name))`. 
The difficulty with versioning is that the dynamic field object can be deleted and then in a later transaction recreated! But as we stated above, the version for an object cannot be reused. This creates a problem when we delete a dynamic field and then recreate it: how do we find a version that is greater than all previous versions? 
Lamport versioning gives us exactly that. Since the dynamic field can only be accessed through its parent, we know that when it was deleted its version was less than or equal to its parent's version. And when it is recreated, we will give it the Lamport version for the transaction. Since that version is increasing for the parent, it must also be greater than any perviously used version for the dynamic field object. 

(Note: there is some subtlety here about dynamic fields on wrapped objects, but the logic works the same. The object containing the wrapped parent can effectively be viewed as the parent of the dynamic field, at least as far as versioning goes)
```

## Move Objects

Move Object versions are calculated using a variant of [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamp) that guarantees that versions never get re-used in all cases:  The new version for objects touched by a transaction is one greater than the max version among all input objects to the transaction.  For example, a transaction transferring an object `O` at version `5` using a gas object `G` at version `3` updates both `O` and `G`'s version to `1 + max(5, 3) = 6`.

The relevance of versions for accessing an object as a transaction input changes depending on that object's ownership:

### Address-owned Objects

Address-owned transaction inputs must be referenced at a specific ID and version.  When a validator signs a transaction with an owned object input at a specific version, that version of the object is **locked** to that transaction, from the validator's perspective: it rejects requests to sign other transactions that require the same input (same ID and version).

> If `F + 1` validators sign one transaction that takes an object as input, and a different `F + 1` validators sign a different transaction that takes the same object as input, that object (and all the other inputs to both transactions) are **equivocated** meaning they cannot be used for any further transactions in that epoch.  This is because neither transaction can form a quorum without relying on a signature from a validator that has already committed the object to a different transaction, which it cannot get.  All locks are reset at the end of the epoch, which frees the objects up once more.

Author: Not sure about removing the indefinite article before `different` here @ronny-mysten -- it reads strangely to me although I don't have the words to explain why.  A simpler example would be:

> 5 people went left and a different 5 people went right

vs

> 5 people went left and different 5 people went right.

> If `F + 1` validators sign one transaction that takes an object as input, and a different `F + 1` validators sign a different transaction that takes the same object as input, that object (and all the other inputs to both transactions) are **equivocated** meaning they cannot be used for any further transactions in that epoch.  This is because neither transaction can form a quorum without relying on a signature from a validator that has already committed the object to a different transaction, which it cannot get.  All locks are reset at the end of the epoch, which frees the objects up once more.

Author: Oh I think I see the confusion now, `F + 1` is not describing the kind of validator, it's the number of validators.

Only an object's owner can equivocate it, but this is not a desirable thing to do.  Equivocation can be avoided by carefully managing the versions of address-owned input objects.

### Immutable Objects

Like Address-owned objects, immutable objects are also referenced at an ID and version, but they do not need to be locked, because their contents and versions do not change.  Their version is relevant because they could have started life as an Address-owned object and later been frozen. The given version identifies the point at which they became immutable.

### Shared Objects

Specifying a shared transaction input is slightly more complex.  It is referenced by its ID, the version it was shared at, and a flag indicating whether it is accessed mutably.  The precise version the transaction will access is **not** specified, because it is decided by consensus during transaction scheduling: When scheduling multiple transactions that touch the same shared object, validators agree the order of those transactions, and pick each transaction's input versions for the shared object's accordingly (one transaction's output version becomes the next transaction's input version, and so on).

> Shared transaction inputs that are referenced immutably participate in scheduling, but don't modify the object or increment its version.

Author: @andll, particularly interested to get your thoughts on read-only shared objects and shared objects in general.

Reviewer 2: This seem reasonable. Just to be clear, the part about "one transaction's output version becomes the next transaction's input version, and so on" is applied to mutable shared references (multiple transactions with immutable references can have same object version as input)

> ### Wrapped Objects

Author: @tnowacki / @gdanezis , curious for your thoughts on wrapped objects and dynamic fields.

The `make_wrapped` function in the example below creates an `Inner` object, wrapped in an `Outer` object which is sent back to the transaction sender.

```
module example::wrapped {
    use sui::object::{Self, UID};
    use sui::transfer;
    use sui::tx_context::{Self, TxContext}
    
    struct Inner has key {
        id: UID,
        x: u64,
    }
    
    struct Outer has key {
        id: UID,
        inner: Inner,
    }
    
    entry fun make_wrapped(ctx: &mut TxContext) {
        transfer::transfer(
            Outer {
                id: object::new(ctx),
                inner: Inner {
                    id: object::new(ctx),
                    x: 42,
                }
            }
            tx_context::sender(ctx),
        )
    }
}
```

> Wrapped object aren't accessible by their ID in the object store, they can only be accessed via the object that wraps them.  In the example above, the owner of `Outer` must specify it as the transaction input and then access its `inner` field to read the instance of `Inner`.  Validators refuse to sign transaction that specify wrapped objects (like the `inner` of an `Outer`) as inputs.  As a result, a wrapped object's version does not need to specified in a transaction that reads it.

Author: Yep, that's right, thanks for clarifying!

Wrapped objects can eventually become "unwrapped", meaning that they are once again accessible at their ID:

```
module example::wrapped {
    // ...
    
    entry fun unwrap(outer: Outer, ctx: &TxContext) {
        let Outer { id, inner } = outer;
        object::delete(id);

>         transfer::transfer(inner, tx_context::sender(ctx))

Author: (it's technically not necessary in this case, but probably a good idea to add it).

    }
}
```

The `unwrap` function above takes an instance of `Outer`, destroys it and sends the `Inner` inside back to the sender.  After calling this function, the previous owner of `Outer` can access `Inner` directly at by its ID, because it has been unwrapped.  An object can be wrapped and unwrapped multiple times across its lifetime, and retains its ID across all those events.

The Lamport timestamp-based versioning schemes helps to ensure that the version that an object is unwrapped at is always greater than the version it was wrapped at, to prevent version re-use:

> - After a transaction, `W` wrapping object `I` in object `O`, `O`'s version is greater than or equal to `I`'s:

Reviewer 3: ```suggestion
- After a transaction, `W` wrapping object `I` in object `O`, `O`'s version is greater than or equal to `I`'s. This means one of the following conditions is true:
```

  - either `I` is an input so has a strictly lower version, 
  - or is new and has an equal version.
- After a later transaction unwrapping `I` out of `O`, we know that 

>   - `O`'s input version is greater than equal to it's version after `W` because it is a later transaction, so the version can only have increased.

Author: Oof, this one was particularly bad on my part 🥲

  - `I`'s version in the output must be strictly greater than `O`'s input version.
  
This leads to the following chain of inequalities:

> - `I`'s version before wrapping

Reviewer 3: ```suggestion

```

> - is less than or equal to `O`'s version after wrapping
> - is less than or equal to `O`'s version before unwrapping
> - is less than `I`'s version after unwrapping

Author: The `is`'s are probably no longer necessary with the new structure.

```suggestion
- less than or equal to `O`'s version after wrapping
- less than or equal to `O`'s version before unwrapping
- less than `I`'s version after unwrapping
```

So `I`'s version before wrapping is less than `I`'s version after unwrapping.

### Dynamic Fields

From a versioning perspective, values held in dynamic fields behave like wrapped objects:

- They are only accessible via their field owner, not as direct transaction inputs.

> - Therefore, their versions do not need to be supplied with the transaction inputs.

Reviewer 3: ```suggestion
- Based on the previous point, you do not need to supply their versions with the transaction inputs.
```

> - Lamport timestamp-based versioning makes sure that when a field is removed and its value becomes accessible by its ID, its version has been incremented.

Reviewer 3: Is it "you" or "transaction" removing the field?

> - Lamport timestamp-based versioning makes sure that when a field is removed and its value becomes accessible by its ID, its version has been incremented.

Author: Either is fine, I think -- your suggestion looks good!

```
(0x17fb7f87e48622257725f584949beac81539a3f4ff864317ad90357c37d82605, 1) => P v1
(0x260f6eeb866c61ab5659f4a89bc0704dd4c51a573c4f4627e40c5bb93d4d500e, 2) => P v2
(0xd24cc3ec3e2877f085bc756337bf73ae6976c38c3d93a0dbaf8004505de980ef, 3) => P v3
```

In the example above, all three versions of the same package are at different IDs, but with increasing versions, and it is possible to call into v1, even though v2 and v3 exist on-chain.

## Framework Packages

Framework packages (such as the Move standard library at `0x1` and the Sui Framework at `0x2`) are a special-case because their IDs must remain stable across upgrades.  The network can upgrade framework packages while preserving their IDs via a system transaction, but can only perform this operation on epoch boundaries because they are considered immutable like other packages.  New versions of framework packages retain the same ID as their predecessor, but increment their version by one:

```
(0x1, 1) => MoveStdlib v1
(0x1, 2) => MoveStdlib v2
(0x1, 3) => MoveStdlib v3
```

> The example above shows the on-chain representation of the first three versions of the Move standard library.

Reviewer 3: ```suggestion
The prior example shows the on-chain representation of the first three versions of the Move standard library.
```
